arch-design-questions.md

```diff
@@ -0,0 +1,93 @@
+# Client design
+
+## Principles
+
+- There are two top level clients - `ServiceBusReceiverClient` and `ServiceBusSenderClient`.
+- Clients are [single-use](#single-use-clients) - they are created with a receive mode, session ID (if applicable), queue/topic-subscription
+  baked in and cannot be changed.
+- The user can choose between two different methods of message delivery - push (via registered message handlers)
+  or pull (via an iterator/reactor pattern or .receiveBatch()-style)
+- Sessions have two different methods for interacting with them:
+  1. [single-use](#single-use-clients)
+  2. [session manager](#sessions-design)
+- Different message types will be used for sending messages vs receiving messages. This is less confusing for
+  the user since we can remove fields that are non-sensical for sending (for instance, `locktoken`).
+
+# Single use clients
+
+Clients are single-use - you cannot change their receive mode or targeted queue/topic/subscription after
+they are created.
+
+This also means that calling .receive() (or registering multiple handlers) isn't a useful activity - the user
+would only be competing against _themselves_ since calling receive() multiple times would just result
+in multiple links, in the same mode, to the same entity, on the same _connection_.
+
+# Sessions design
+
+The primary use case for non-session bus users when receiving is opening
+up a single client instance, connecting to a targeted queue (or topic/subscription)
+and receiving from it.
+
+In this case the lifetime of the client is essentially the lifetime of their application which makes
+connection management a non-issue.
+
+Sessions differ in that the user's application is not tied to a single session - the user
+is expected to open up and read from many sessions during the lifetime of their application.
+
+This requires a few design considerations:
+
+## Connection "explosion"
+
+**Problem**
```

Comment:
I think that its worth calling out that the service team has a goal of allowing transparency for customers with respect to connections, as it is something at the service level constrained by a quota.   Their model attempts to equate 1 client == 1 connection.

There have been cases of customers requesting to share connections between clients as a feature for both Service Bus and Event Hubs , which is another benefit to using this abstraction and allowing customers visibility and control over it.

Reply:
I've added a section on connection sharing to address this as it turned out to be both a session concern and a non-session concern. Let me know if I captured everything since we kind of want to hibernate the discussion for a few weeks. :)